### PR TITLE
Replace Tinymath with math.js

### PR DESCRIPTION
**Pull request by gulderov**
### Description
Migrate from tinymath to math.js

[Tinymath](https://github.com/elastic/tinymath) has been deprecated for several years, and as part of the modernization plan (https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4306), we need to replace it.

[Math.js](https://github.com/josdejong/mathjs) is an extensive math library for JavaScript and Node.js. It features a flexible expression parser with support for symbolic computation, comes with a large set of built-in functions and constants. It is maintained project with good [docs](https://mathjs.org/docs/reference/functions.html).

The [evaluate](https://mathjs.org/docs/reference/functions/evaluate.html) method can replace tinymath.evaluate with [security](https://mathjs.org/docs/expressions/security.html) in mind

Examples :
`2 + 3`
`add(params.count, 3)`
`[a = params.count * 100, f(x) = sqrt(x), f(a)]`

Examples that would NOT work: 

`parse`, `derivative` and `evaluate` are disabled as risky:
`[
h = parse('x^2 + x'),
dh = derivative(h, 'x'),
dh.evaluate({ x: params.count })
]`

### Issues Resolved

closes https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3655
closes https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3656
closes https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3657

## Testing the changes
Open [eCommerce] Revenue Dashboard with sample data => Edit => Create new => New Visualization TSVB => Press + `Add Series` => Press + `Add Metric` => Last Aggregation `Math` => Variable name: `count`, Expression: `params.count + 10` => Check graph

Screenshot 2023-07-04 at 12 58 27
Screenshot 2023-07-04 at 12 58 52
Screenshot 2023-07-04 at 12 59 06

### Check List
- [x] `yarn test:jest`
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/series/evaluate.ts
Comment (abbyhu2000) [inline src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/series/evaluate.ts:1]: Do we disable the following math.js fu?nctions due to security issues? How do you know these are the only functions that we need to disable?

File: src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/series/evaluate.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+math.import(
+  {
+    import() {
+      throw new Error(MATH_THROW_MSG);
+    },
+    createUnit() {
+      throw new Error(MATH_THROW_MSG);
+    },
+    evaluate() {
```
Comment (abbyhu2000) [inline src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/series/evaluate.ts:20]: Is this evaluate() function different than `math.evaluate()` from line 9?

**Review by abbyhu2000 — COMMENTED** [review]
Thanks @gulderov, no issues functionality-wise, but just some small questions.

**Comment by AMoo-Miki** [comment]
Thanks for this PR.

> parse, derivative and evaluate are disabled

I labelled this as `breaking change`; if that is not what you meant and the new library can be used just like the old one, lemme know and I will remove the label.

File: src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/series/evaluate.ts
Comment (gulderov) [reply src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/series/evaluate.ts:1]: These functions are marked as unsafe by the mathjs documentation https://mathjs.org/docs/expressions/security.html


A user could try to inject malicious JavaScript code via the expression parser. The expression parser of mathjs offers a sandboxed environment to execute expressions which should make this impossible. It’s possible though that there are unknown security vulnerabilities, so it’s important to be careful, especially when allowing server side execution of arbitrary expressions.

The expression parser of mathjs parses the input in a controlled way into an expression tree or abstract syntax tree (AST). In a “compile” step, it does as much as possible preprocessing on the static parts of the expression, and creates a fast performing function which can be used to evaluate the expression repeatedly using a dynamically passed scope.

The parser actively prevents access to JavaScripts internal eval and new Function which are the main cause of security attacks. Mathjs versions 4 and newer does not use JavaScript’s eval under the hood.

File: src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/series/evaluate.ts (hunk 1 above)
Comment (gulderov) [reply src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/series/evaluate.ts:20]: It is the same `evaluate` function. The idea is to disable risky functions on the user side. We can find a list of all risky functions in the `mathjs` documentation: https://mathjs.org/docs/expressions/security.html.

So it is possible to do: `mathjs.evaluate('add(2, 2)')`.
At the same time, it is impossible to evaluate inside the evaluate function itself: `mathjs.evaluate('evaluate("add(2,2)")')`.

**Comment by gulderov** [comment]
> Thanks @gulderov, no issues functionality-wise, but some questions regarding the disabled functions.

@abbyhu2000 Thank you so much! 
I have answered all the questions. Please let me know if I need to provide more information. Feel free to mark conversations as resolved in other case

**Comment by gulderov** [comment]
> Thanks for this PR.
> 
> > parse, derivative and evaluate are disabled
> 
> I labelled this as `breaking change`; if that is not what you meant and the new library can be used just like the old one, lemme know and I will remove the label.

@AMoo-Miki, library change could be marked as breaking for sure
As I can see `tinymath` is more limited than `mathjs` and disabled functions are not possible in `tinymath`

It is not a drop in replacement for users. But in most cases it would work as is or with minimal changes
* clamp -> `['x=1', 'minimum=2', 'maximum=3', 'min(max(x, minimum), maximum)']` // 2
* first -> `['a=[1,2,3]', 'a[1]']` // 1
* last -> `['a=[1,2,3]', 'a[a.size()]']` // 3
* unique -> `['setDistinct([1,1,1,2])']` // [1, 2]

**Review by abbyhu2000 — DISMISSED** [review]
Thanks for the explanations!

**Review by ananzh — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]

**Comment by ananzh** [comment]
cypress fail but not caused by this PR.

**Pull request merged**